### PR TITLE
Fix problem with unbound type parameters as extension targets

### LINKED ISSUE #2113: Unhandled exceptions with --link-to-remote (type 'DynamicTypeImpl' is not a subtype of type 'InterfaceType')
Similar to #2101, I've found another exception that is triggered when running dartdoc with `--link-to-remote`:

https://pub.dev/packages/aws_sns_api

On pub.dev:
https://pub.dev/documentation/aws_sns_api/latest/log.txt
````
type 'DynamicTypeImpl' is not a subtype of type 'InterfaceType'
#0      TypeParameterElementType.interfaceType (package:dartdoc/src/element_type.dart:261:38)
#1      DefinedElementType.instantiatedType (package:dartdoc/src/element_type.dart:331:12)
#2      Extension._couldApplyTo (package:dartdoc/src/model.dart:1350:47)
#3      Extension.couldApplyTo (package:dartdoc/src/model.dart:1346:33)
#4      Class.potentiallyApplicableExtensions.<anonymous closure> (package:dartdoc/src/model.dart:805:27)
````

The fix to #2101 may have been fixed this one too, please double-check...

### PR DESCRIPTION
**Pull request by jcollins-g**
Fixes #2113.  [edited]

We didn't handle unbound type parameters as an extension target, and this highlighted a previously known bug in our handling of instantiated types.

Rationalize `ElementType` so that all `ElementType`s can do subtype checks.  We also catch up with the latest changes to analyzer and avoid use of `type.name`.  I feel the type checking code is now easier to rationalize about and is more correct.

### CODE REVIEW

**Comment by jcollins-g** [comment]
This PR has serious bugs; holding off on reviews until they get straightened out.

File: lib/src/element_type.dart [hunk 1]
```diff
@@ -71,12 +71,17 @@ abstract class ElementType extends Privacy {
 
   String get linkedName;
 
-  String get name => type.name ?? type.element.name;
+  String get name => type.element.name;
```
Comment (scheglov) [inline lib/src/element_type.dart]: FYI, `element` might be `null`, e.g. for `FunctionType`.

File: lib/src/element_type.dart [hunk 2]
```diff
@@ -274,32 +292,38 @@ abstract class DefinedElementType extends ElementType {
...
   DartType get instantiatedType {
     if (_instantiatedType == null) {
-      if (!_interfaceType.typeArguments.every((t) => t is InterfaceType)) {
+      if (_bound is InterfaceType &&
+          !(_bound as InterfaceType)
+              .typeArguments
+              .every((t) => t is InterfaceType)) {
         var typeSystem = library.element.typeSystem as TypeSystemImpl;
-        _instantiatedType = typeSystem.instantiateToBounds(_interfaceType);
+        _instantiatedType = typeSystem.instantiateToBounds(_bound);
```
Comment (scheglov) [inline lib/src/element_type.dart]: Please use `ClassElement.instantiateToBounds` if it is available in your version of analyzer.

File: lib/src/element_type.dart (hunk 1 above)
Comment (jcollins-g) [reply lib/src/element_type.dart]: Moved implementation to where an element is guaranteed.

File: lib/src/element_type.dart (hunk 2 above)
Comment (jcollins-g) [reply lib/src/element_type.dart]: Will migrate in a separate PR.

**Pull request merged**